router/core/graphql_prehandler.go

```diff
@@ -167,28 +176,92 @@ func (h *PreHandler) Handler(next http.Handler) http.Handler {
 			metrics.Finish(finalErr, statusCode, writtenBytes)
 		}()
 
+		var body []byte
+		var files []httpclient.File
+		var err error
 		// XXX: This buffer needs to be returned to the pool only
 		// AFTER we're done with body (retrieved from parser.ReadBody())
 		buf := pool.GetBytesBuffer()
 		defer pool.PutBytesBuffer(buf)
+		if r.Header.Get("Content-Type") == "" || r.Header.Get("Content-Type") == "application/json" {
+			body, err = h.operationProcessor.ReadBody(buf, r.Body)
+			if err != nil {
+				finalErr = err
 
-		body, err := h.operationProcessor.ReadBody(buf, r.Body)
-		if err != nil {
-			finalErr = err
+				// This error is expected e.g. when the client defines (Content-Length) and aborts the request before
+				// It means that EOF was encountered in the middle of reading the body. This is not a server error.
+				if errors.Is(err, io.ErrUnexpectedEOF) {
+					requestLogger.Debug("unexpected EOF while reading request body", zap.Error(err))
+				} else {
+					requestLogger.Error("failed to read request body", zap.Error(err))
+				}
 
-			// This error is expected e.g. when the client defines (Content-Length) and aborts the request before
-			// It means that EOF was encountered in the middle of reading the body. This is not a server error.
-			if errors.Is(err, io.ErrUnexpectedEOF) {
-				requestLogger.Debug("unexpected EOF while reading request body", zap.Error(err))
-			} else {
-				requestLogger.Error("failed to read request body", zap.Error(err))
+				writeOperationError(r, w, requestLogger, err)
+				return
+			}
+		} else if strings.Contains(r.Header.Get("Content-Type"), "multipart/form-data") {
+			tempBuf := pool.GetBytesBuffer()
+			defer pool.PutBytesBuffer(tempBuf)
+			if _, err = h.operationProcessor.ReadBody(tempBuf, r.Body); err != nil {
+				finalErr = err
+				writeOperationError(r, w, requestLogger, err)
+				return
+			}
+			r.Body = io.NopCloser(bytes.NewReader(tempBuf.Bytes()))
+
+			parser, err := httpform.NewParser(r)
+			if err != nil {
+				writeOperationError(r, w, requestLogger, err)
+				return
+			}
+
+			err = parser.Register("operations", func(reader io.Reader, header formstream.Header) error {
+				body, err = h.operationProcessor.ReadBody(buf, reader)
+				if err != nil {
+					return err
+				}
+				return nil
+			}, formstream.WithRequiredPart("operations"), formstream.WithRequiredPart("map"))
+			if err != nil {
+				writeOperationError(r, w, requestLogger, err)
+				return
+			}
+
+			// We will register a handler for each file in the request. AFAIK, we can't know how many files we have
+			// before parsing the request, so we will support 10 files max.
+			for i := 0; i < h.maxUploadFiles; i++ {
```

Comment:
When the limit is hit we should return error otherwise it gets swallowed and is hard to debug.

Reply:
Done